internal/provider/resource_model.go

```diff
@@ -4,326 +4,500 @@ import (
+	// Prevent panic if the provider has not been configured.
+	if r.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "model", "import")
+		return
+	}
+
+	// Import command takes the model name as an argument
+	// `terraform import juju_model.RESOURCE_NAME MODEL_NAME`
+	modelName := req.ID
+
+	// We set the ID to the modelUUID here
+	modelInfo, err := r.client.Models.GetModelByName(modelName)
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to find model, got error: %s", err))
+		return
+	}
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("name"), modelInfo.Name)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("id"), modelInfo.UUID)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	r.trace(fmt.Sprintf("Imported model resource: %v", modelInfo.Name))
+}
+
+func (r *modelResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	// Prevent panic if the provider has not been configured.
+	if r.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "model", "create")
+		return
+	}
+
+	var plan modelResourceModel
+
+	// Read Terraform configuration from the request into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
-	name := d.Get("name").(string)
-	cloud := d.Get("cloud").([]interface{})
-	config := d.Get("config").(map[string]interface{})
-	credential := d.Get("credential").(string)
-	readConstraints := d.Get("constraints").(string)
+	// Acquire modelName, clouds, config, credential & constraints from the model plan
+	modelName := plan.Name.ValueString()
+	var clouds []nestedCloud
+	resp.Diagnostics.Append(plan.Cloud.ElementsAs(ctx, &clouds, false)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	var config map[string]string
+	resp.Diagnostics.Append(plan.Config.ElementsAs(ctx, &config, false)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	credential := plan.Credential.ValueString()
+	readConstraints := plan.Constraints.ValueString()
 
-	var parsedConstraints constraints.Value = constraints.Value{}
+	parsedConstraints := constraints.Value{}
 	var err error
 	if readConstraints != "" {
+		// TODO (cderici): this may be moved into internal/model so
+		// resource_model can avoid importing juju/core/constraints
 		parsedConstraints, err = constraints.Parse(readConstraints)
 		if err != nil {
-			return diag.FromErr(err)
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to parse constraints, got error: %s", err))
+			return
 		}
 	}
 
-	response, err := client.Models.CreateModel(juju.CreateModelInput{
-		Name:        name,
-		CloudList:   cloud,
+	cloudNameInput := clouds[0].Name.ValueString()
+	cloudRegionInput := clouds[0].Region.ValueString()
```

Comment:
You're making an assumption that the cloud block exists, which it's optional. Causes a panic with
```
terraform {
  required_providers {
    juju = {
      version = ">= 0.7.0"
      source  = "juju/juju"
    }
  }
}

provider "juju" {
}

resource "juju_model" "testmodel" {
  name = "machine-test"
}

```